File: spx-analyst/src/web/chat_api.py
"""FastAPI routes for the research assistant chat API."""

from __future__ import annotations

import json
from typing import Iterator

from fastapi import APIRouter, HTTPException
from fastapi.responses import StreamingResponse

from ..chat_service import ChatServiceError, SessionNotFoundError, get_chat_service
from ..chat_sessions import get_session
from ..openai_assistant import ThreadMessage
from ..schemas import ChatSessionRecord
from .models import (
    ChatMessageResponse,
    ChatSessionResponse,
    CreateChatSessionRequest,
    PostChatMessageRequest,
    RenameChatSessionRequest,
)

router = APIRouter(prefix="/api/chat", tags=["chat"])


def _session_response(record: ChatSessionRecord) -> ChatSessionResponse:
    return ChatSessionResponse(
        id=record.id,
        title=record.title,
        created_at=record.created_at,
        updated_at=record.updated_at,
    )


def _message_response(message: ThreadMessage) -> ChatMessageResponse:
    return ChatMessageResponse(
        id=message.id,
        role=message.role,  # type: ignore[arg-type]
        content=message.content,
        created_at=message.created_at,
    )


@router.get("/sessions", response_model=list[ChatSessionResponse])
def list_chat_sessions() -> list[ChatSessionResponse]:
    service = get_chat_service()
    return [_session_response(s) for s in service.list_sessions()]


@router.post("/sessions", response_model=ChatSessionResponse, status_code=201)
def create_chat_session(body: CreateChatSessionRequest) -> ChatSessionResponse:
    service = get_chat_service()
    try:
        record = service.create_session(title=body.title)
    except ChatServiceError as exc:
        raise HTTPException(status_code=503, detail=str(exc)) from exc
    return _session_response(record)


@router.get("/sessions/{session_id}/messages", response_model=list[ChatMessageResponse])
def get_chat_messages(session_id: str) -> list[ChatMessageResponse]:
    service = get_chat_service()
    try:
        messages = service.get_messages(session_id)
    except SessionNotFoundError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    except ChatServiceError as exc:
        raise HTTPException(status_code=503, detail=str(exc)) from exc
    return [_message_response(m) for m in messages]


@router.patch("/sessions/{session_id}", response_model=ChatSessionResponse)
def rename_chat_session(session_id: str, body: RenameChatSessionRequest) -> ChatSessionResponse:
    service = get_chat_service()
    try:
        record = service.rename_session(session_id, body.title)
    except SessionNotFoundError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    except ChatServiceError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    return _session_response(record)


@router.delete("/sessions/{session_id}", status_code=204)
def delete_chat_session(session_id: str) -> None:
    service = get_chat_service()
    try:
        service.delete_session(session_id)
    except SessionNotFoundError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc


@router.post("/sessions/{session_id}/messages")
def post_chat_message(session_id: str, body: PostChatMessageRequest) -> StreamingResponse:
    service = get_chat_service()
    content = body.content.strip()
    if not content:
        raise HTTPException(status_code=400, detail="message must not be empty")

    try:
        get_session(session_id, service.settings)
    except SessionNotFoundError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc

    def event_stream() -> Iterator[str]:
        try:
            for chunk in service.stream_reply(session_id, content):
                payload = json.dumps({"text": chunk}, ensure_ascii=False)
                yield f"data: {payload}\n\n"
            yield "data: [DONE]\n\n"
        except ChatServiceError as exc:
            payload = json.dumps({"error": str(exc)}, ensure_ascii=False)
            yield f"data: {payload}\n\n"

    return StreamingResponse(event_stream(), media_type="text/event-stream")


def clear_chat_service_cache() -> None:
    """Test helper — reset cached ChatService singleton."""
    get_chat_service.cache_clear()

File: spx-analyst/web/lib/chat-api.ts
import { ApiError, type ChatMessage, type ChatSession } from "@/lib/types";

function apiBase(): string {
  if (typeof window === "undefined") {
    return process.env.API_BASE_URL ?? "http://127.0.0.1:8000";
  }
  return "";
}

function parseErrorDetail(raw: string): string {
  const trimmed = raw.trim();
  if (!trimmed) {
    return "Request failed";
  }
  try {
    const payload = JSON.parse(trimmed) as { detail?: string | { msg?: string }[] };
    if (typeof payload.detail === "string") {
      return payload.detail;
    }
    if (Array.isArray(payload.detail) && payload.detail[0]?.msg) {
      return payload.detail[0].msg;
    }
  } catch {
    // Not JSON — use raw text.
  }
  return trimmed;
}

async function fetchJson<T>(path: string, init?: RequestInit): Promise<T> {
  const response = await fetch(`${apiBase()}${path}`, {
    cache: "no-store",
    ...init,
    headers: {
      "Content-Type": "application/json",
      ...(init?.headers ?? {}),
    },
  });

  if (!response.ok) {
    const detail = parseErrorDetail(await response.text());
    throw new ApiError(detail || response.statusText, response.status);
  }

  if (response.status === 204) {
    return undefined as T;
  }

  return response.json() as Promise<T>;
}

export async function listChatSessions(): Promise<ChatSession[]> {
  return fetchJson<ChatSession[]>("/api/chat/sessions");
}

export async function createChatSession(title = "New conversation"): Promise<ChatSession> {
  return fetchJson<ChatSession>("/api/chat/sessions", {
    method: "POST",
    body: JSON.stringify({ title }),
  });
}

export async function renameChatSession(
  sessionId: string,
  title: string,
): Promise<ChatSession> {
  return fetchJson<ChatSession>(`/api/chat/sessions/${sessionId}`, {
    method: "PATCH",
    body: JSON.stringify({ title }),
  });
}

export async function deleteChatSession(sessionId: string): Promise<void> {
  await fetchJson<void>(`/api/chat/sessions/${sessionId}`, { method: "DELETE" });
}

export async function getChatMessages(sessionId: string): Promise<ChatMessage[]> {
  return fetchJson<ChatMessage[]>(`/api/chat/sessions/${sessionId}/messages`);
}

export interface StreamChatHandlers {
  onChunk: (text: string) => void;
  onError: (message: string) => void;
  onDone: () => void;
}

export async function streamChatMessage(
  sessionId: string,
  content: string,
  handlers: StreamChatHandlers,
): Promise<void> {
  const response = await fetch(`${apiBase()}/api/chat/sessions/${sessionId}/messages`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ content }),
  });

  if (!response.ok) {
    const detail = parseErrorDetail(await response.text());
    throw new ApiError(detail || response.statusText, response.status);
  }

  if (!response.body) {
    throw new ApiError("Empty response body", response.status);
  }

  const reader = response.body.getReader();
  const decoder = new TextDecoder();
  let buffer = "";

  while (true) {
    const { done, value } = await reader.read();
    if (done) {
      break;
    }

    buffer += decoder.decode(value, { stream: true });
    const lines = buffer.split("\n");
    buffer = lines.pop() ?? "";

    for (const line of lines) {
      if (!line.startsWith("data: ")) {
        continue;
      }
      const data = line.slice(6).trim();
      if (data === "[DONE]") {
        handlers.onDone();
        return;
      }
      try {
        const payload = JSON.parse(data) as { text?: string; error?: string };
        if (payload.error) {
          handlers.onError(payload.error);
          throw new ApiError(payload.error, 502);
        }
        if (payload.text) {
          handlers.onChunk(payload.text);
        }
      } catch (error) {
        if (error instanceof ApiError) {
          throw error;
        }
        handlers.onError("Invalid stream payload");
        throw new ApiError("Invalid stream payload", 502);
      }
    }
  }

  handlers.onDone();
}
